Objects/dictobject.c

```diff
@@ -1650,6 +1650,69 @@ PyDict_SetItem(PyObject *op, PyObject *key, PyObject *value)
     return insertdict(mp, key, hash, value);
 }
 
+/* Initialize dict from keywords tuple and values array.
+ *
+ * - Dict must be empty before first call, presized, and combined.
+ * - keys must be distinct.
+ */
+static int
+dict_init_from_items(PyObject *op, PyObject *keys, PyObject * const *values)
+{
+    assert(PyDict_Check(op));
+    assert(PyTuple_Check(keys));
+
+    PyDictObject *mp = (PyDictObject *)op;
+    assert(!_PyDict_HasSplitTable(mp));
+
+    Py_ssize_t size = PyTuple_GET_SIZE(keys);
+    assert(mp->ma_used == 0);
+    assert(mp->ma_keys->dk_usable >= size);
+
+    PyDictKeyEntry *ep = &DK_ENTRIES(mp->ma_keys)[0];
+
+    int all_unicode = 1;
+
+    for (Py_ssize_t i = 0; i < size; i++) {
+        PyObject *key = PyTuple_GET_ITEM(keys, i);
+        Py_INCREF(key);
+        if (!PyUnicode_CheckExact(key)) {
+            all_unicode = 0;
+        }
+
+        Py_hash_t hash;
+        if (!PyUnicode_CheckExact(key) ||
+            (hash = ((PyASCIIObject *) key)->hash) == -1)
+        {
+            hash = PyObject_Hash(key);
+            if (hash == -1) {
+                Py_DECREF(key);
+                return -1;
+            }
+        }
+
+        Py_INCREF(values[i]);
+        ep[i].me_key = key;
+        ep[i].me_hash = hash;
+        ep[i].me_value = values[i];
+        mp->ma_keys->dk_nentries++;
+    }
+
+    build_indices(mp->ma_keys, ep, size);
+
+    if (!all_unicode) {
+        mp->ma_keys->dk_lookup = lookdict;
+    }
+    mp->ma_keys->dk_usable -= size;
+    mp->ma_used = size;
+    mp->ma_version_tag = DICT_NEXT_VERSION();
+
+    for (Py_ssize_t i = 0; i < size; i++) {
+        MAINTAIN_TRACKING(mp, ep[i].me_key, ep[i].me_value);
```

Comment:
Can't be inside the previous loop?

Reply:
Note that `PyObject_Hash()` may release GIL.
Once we track the object, GC will scan this object even though its under construction. And this object can be exposed by `gc` module.

I prefer keep it hidden until dict is fully constructed.